zipkin/src/main/java/zipkin/internal/Node.java

```diff
@@ -145,9 +145,9 @@ public void addNode(Long parentId, long id, @Nullable V value) {
       for (Map.Entry<Long, Long> entry : idToParent.entrySet()) {
         Node<V> node = idToNode.get(entry.getKey());
         Node<V> parent = idToNode.get(entry.getValue());
-        if (parent == null && rootNode == null) { // handle headless trace
-          rootNode = node;
-        } else if (parent == null) { // attribute missing parents to root
+        if (parent == null) { // handle headless trace
+          if (rootNode == null)
```

Comment:
rootNode = Node.MISSING_ROOT

^^ something like this makes the code more obvious, plus you can search references for Node.MISSING_ROOT, where you can't search references for null